features/FEATURE_BLE/ble/gap/AdvertisingDataBuilder.h

```diff
@@ -0,0 +1,552 @@
+     *
+     * @param buffer Pointer to buffer to be used for storing advertising data.
+     * @param buffer_size Size of the buffer.
+     * @note Use Gap::getMaxAdvertisingDataLength() to find out how much can be accepted.
+     */
+    AdvertisingDataBuilder(uint8_t* buffer, size_t buffer_size) :
+        _buffer(buffer, buffer_size),
+        _payloadLen(0) {
+    }
+
+    /**
+     * Get the subspan of the buffer containing valid data.
+     *
+     * @return A Span containing the payload.
+     */
+    mbed::Span<uint8_t> getAdvertisingData() const {
+        return _buffer.first(_payloadLen);
+    }
+
+    /**
+     * Adds a new field into the payload.
+     *
+     * If the supplied advertising data type is already present in the
+     * advertising payload, then the value is updated.
+     *
+     * @param[in] advDataType The type of the field to add.
+     * @param[in] payload Pointer to the value of the field to add.
+     * @param[in] len Size in bytes of the value to add.
+     *
+     * @return BLE_ERROR_NONE on success.
+     * @return BLE_ERROR_BUFFER_OVERFLOW if the new value causes the advertising
+     * buffer to overflow.
+     *
+     * @note When the specified data type is INCOMPLETE_LIST_16BIT_SERVICE_IDS,
+     * COMPLETE_LIST_16BIT_SERVICE_IDS, INCOMPLETE_LIST_32BIT_SERVICE_IDS,
+     * COMPLETE_LIST_32BIT_SERVICE_IDS, INCOMPLETE_LIST_128BIT_SERVICE_IDS,
+     * COMPLETE_LIST_128BIT_SERVICE_IDS or LIST_128BIT_SOLICITATION_IDS, the
+     * supplied value is appended to the values present in the payload.
+     */
+    ble_error_t addData(GapAdvertisingConstant::DataType advDataType, const uint8_t *payload, uint8_t len)
+    {
+        return appendField(advDataType, payload, len);
+    }
+
+    /**
+     * Clears the advertising data payload.
+     *
+     * @post getPayloadLen() returns 0.
+     */
+    void clear(void)
+    {
+        memset(_buffer.data(), 0, _buffer.size());
+        _payloadLen = 0;
+    }
+
+private:
+    /**
+     * Append advertising data based on the specified type.
+     *
+     * @param[in] advDataType Type of the new data.
+     * @param[in] payload Pointer to the data to be appended to the advertising
+     * payload.
+     * @param[in] len Length of the data pointed to by @p payload.
+     *
+     * @return BLE_ERROR_NONE on success.
+     * @return BLE_ERROR_BUFFER_OVERFLOW if the specified data would cause the
+     * advertising buffer to overflow.
+     */
+    ble_error_t appendField(GapAdvertisingConstant::DataType advDataType, const uint8_t *payload, uint8_t len)
+    {
+        /* Make sure we don't exceed the buffer size */
+        if (_payloadLen + len + 2 > _buffer.size()) {
+            return BLE_ERROR_BUFFER_OVERFLOW;
+        }
+
+        /* Field length. */
+        memset(_buffer.data() + _payloadLen, len + 1, 1);
+        _payloadLen++;
+
+        /* Field ID. */
+        memset(_buffer.data() + _payloadLen, (uint8_t)advDataType, 1);
```

Comment:
Offset issue?

Reply:
Not really, the payload length is incremented by one one line above.